### ISSUE TITLE
reuse connections on table.insert

### ISSUE THREAD

**Issue by alexander-fenster** [googleapis/nodejs-bigquery#41]
TL;DR: the current implementation of BigQuery establishes a new connections on each table.insert() call. This causes problems for Cloud Functions because of connection quota that our users hit. Please update the client libraries to reuse connections.

Why new connections cause problems:
----------------------------------------------------
Cloud Functions is a serverless platform for executing snippets of code. Many our customers use function to store events in Big Query. GCF has a quota on connections, and this quota is relatively low because each connection requires a NAT port to be allocated and maintained until TCP packet timeout; and there are only 32k ports per server.

Why BigQuery is worse than other client libraries:
-----------------------------------------------------------------
I found three types of clients:
1. Clients that always reuse connections - even if we construct the client object in a local scope - @google-cloud/storage
2. Clients that reuse connectoins only if we declare the client object at global scope - @google-cloud/pubsub, @google-cloud/language, @google-cloud/spanner
3. Clients that never reuse connectoins - @google-cloud/bigquery

It would be nice if all libraries worked as 1.

How to reproduce the problem:
-----------------------------------------
Every call to the function exported by the code below creates a new connection.

******************* function.js ****************
```js
const bigquery = require('@google-cloud/bigquery')();
const table = bigquery.dataset('Tests').table('RandomNumbers');

exports.function = function(req, res) {
  var r = Math.floor(Math.random() * 100);
  console.log("Generated random number " + r);
  var row = {number: r};
  table.insert(row, function(err, apiResponse) {
    if (err) {
      result = 'Error inserting data to bigquery';
      console.log(result + ": " + JSON.stringify(err, null, 2));
      res.status(500).send("Error" + err);
    } else {
      res.status(200).send("" + r);
    }
  });
};
```
*********************************************

************* package.json ****************
```json
{
  "dependencies": {
    "@google-cloud/bigquery": "^0.10.0"
  }
}
```
*********************************************

We also found that BigQuery uses IPv4, while other libraries prefer IPv6. This should also be fixed.

[Googlers: internal tracking id b/68240537]

**Comment by stephenplusplus** [comment]
I believe what we want is a `Keep-Alive` header for our requests. This is possible using the npm `request` module that we use, and is actually the default behavior. However, we [turned it off](https://github.com/GoogleCloudPlatform/google-cloud-node/pull/2282) specifically for Google Cloud Functions due to users receiving lots of socket hang up errors (see here and the rest of the discussion in the issue: https://github.com/GoogleCloudPlatform/google-cloud-node/issues/2254#issuecomment-299609092).

**Comment by stephenplusplus** [comment]
Related: [Storage API has poor performance in Google Cloud Functions](https://github.com/googleapis/nodejs-storage/issues/101)

**Comment by shannb** [comment]
Any resolution to this?  @alexander-fenster  were you able to resolve the issue but setting the forever agent to false on a fork of google-cloud/storage?

**Comment by stephenplusplus** [comment]
If we confirm that [this PR](https://github.com/GoogleCloudPlatform/google-cloud-node/pull/2282) was a mistake, we would disable it in the same place, and all APIs would be immediately corrected.

For any users that want to undo the change from that PR, you can use request overrides to turn Keep-Alive back on:

```js
const BigQuery = require('@google-cloud/bigquery')
const bigquery = new BigQuery({...})

bigquery.interceptors.push({
  request: reqOpts => {
    reqOpts.forever = true
    return reqOpts
  }
})
```

**Comment by danoscarmike** [comment]
Ping.  Next steps?

**Comment by emilyplusplus** [comment]
Any updates? @stephenplusplus

**Comment by stephenplusplus** [comment]
Sorry, I was hoping @alexander-fenster would advise.

**Comment by amineharbaoui** [comment]
Any news for this issue ??

**Comment by callmehiphop** [comment]
@stephenplusplus @alexander-fenster is this still an issue? Haven't seen any chatter about it in a long while

**Comment by JustinBeckwith** [comment]
Just to be safe, it would be great to run a quickie system test that does a bunch of inserts back-to-back, and monitors the number of open network connections.  I have no idea how the move to `teeny-request` could have affected things here tbh

**Comment by shannb** [comment]
We have resolved the issues we had with bigquery inserts.   We haven't had
any since then.

**Comment by sduskis** [comment]
@callmehiphop, can you please add a system test?

**Comment by shannb** [comment]
This is resolved.  No need to follow up any further.

**Comment by romaincointepas** [comment]
Bumping, as connections do not seem to be reused anymore (anywhere, not only in Cloud functions).

I'm wondering if the switch from [`request`](https://www.npmjs.com/package/request) to [`teeny-request`](https://www.npmjs.com/package/teeny-request) could be the culprit as `teeny-request` does not support the `forever` and `pool` options (`request` did), and the default behavior of node HTTP(S) agents is to not send the `Connection: keep-alive` header.

---

Some digging:

[`@google-cloud/bigquery`](https://www.npmjs.com/package/@google-cloud/bigquery) relies on [`@google-cloud/common`](https://www.npmjs.com/package/@google-cloud/common) for requests, with the following defaults:
```ts
const requestDefaults = {
  timeout: 60000,
  gzip: true,
  forever: true,
  pool: {
    maxSockets: Infinity,
  },
};
```
https://github.com/googleapis/nodejs-common/blob/master/src/util.ts#L35-L42

[`@google-cloud/common`](https://www.npmjs.com/package/@google-cloud/common) then uses [`teeny-request`](https://www.npmjs.com/browse/depended/teeny-request) for requests that itself uses [`node-fetch`](https://www.npmjs.com/package/node-fetch).

Both the `forever` and the `pool` option seems to be ignored (not converted then passed as option to `node-fetch`):
```ts
function requestToFetchOptions(reqOpts: r.Options) {
  const options: f.RequestInit = {
    method: reqOpts.method || 'GET',
    ...(reqOpts.timeout && {timeout: reqOpts.timeout}),
    ...(reqOpts.gzip && {compress: reqOpts.gzip}),
  };

  if (typeof reqOpts.json === 'object') {
    // Add Content-type: application/json header
    reqOpts.headers = reqOpts.headers || {};
    reqOpts.headers['Content-Type'] = 'application/json';

    // Set body to JSON representation of value
    options.body = JSON.stringify(reqOpts.json);
  } else {
    if (typeof reqOpts.body !== 'string') {
      options.body = JSON.stringify(reqOpts.body);
    } else {
      options.body = reqOpts.body;
    }
  }

  options.headers = reqOpts.headers as Headers;

  let uri = ((reqOpts as r.OptionsWithUri).uri ||
    (reqOpts as r.OptionsWithUrl).url) as string;
  if (reqOpts.useQuerystring === true || typeof reqOpts.qs === 'object') {
    const qs = require('querystring');
    const params = qs.stringify(reqOpts.qs);
    uri = uri + '?' + params;
  }

  if (reqOpts.proxy || process.env.HTTP_PROXY || process.env.HTTPS_PROXY) {
    const proxy = (process.env.HTTP_PROXY || process.env.HTTPS_PROXY)!;
    options.agent = new HttpsProxyAgent(proxy);
  }

  return {uri, options};
}
```
https://github.com/googleapis/teeny-request/blob/master/src/index.ts#L22-L60

node HTTP(S) agents default to not reusing sockets:
  - `keepAlive` defaults to `false`
  - `maxSockets` defaults to `Infinity`
  - `The Connection: keep-alive header is always sent when using an agent except when the Connection header is explicitly specified or when the keepAlive and maxSockets options are respectively set to false and Infinity, in which case Connection: close will be used.` (https://nodejs.org/api/http.html#http_new_agent_options)
